Moya/Plugin.swift

```diff
@@ -0,0 +1,44 @@
+import Foundation
+import Alamofire
+
+/// A Moya Plugin receives callbacks to perform side effects wherever a request is sent or received.
+///
+/// for example, a plugin may be used to
+///     - log network requests
+///     - hide and show a network avtivity indicator 
+///     - inject additional information into a request
+public class Plugin<Target: MoyaTarget> {
+    
+    // NOTE:
+    //
+    // We cannot implement `Plugin` as a generic protocol here, because `MoyaProvider` needs
+    // to keep references to an array of plugins and a generic protocol cannot be used as an array constraint.
+    //
+    // i.e.
+    // protocol Plugin {
+    //      typealias T: MoyaTarget
+    //      ...
+    // }
+    //
+    // let plugins = [Plugin]()
+    // 
+    // This does not work, because `plugins` is now unable to infer the actual type of the typealias `T`.
+    
+    func willSendRequest(token: Target, request: Alamofire.Request) -> Alamofire.Request {
```

Comment:
Good point. Alamofire.Request is passed by reference anyway, maybe we can get rid of the return anyway?

In case we want to go forward with the overloading, calling just  `plugins.forEach{ request = $0.willSendRequest(token, request: request) }` would no longer work, since it will always call the function with the return value.

Should we invoke both methods within the `forEach` closure?